### PR TITLE
[system-probe] Add HTTP monitoring userspace side

### PR DESCRIPTION
**Pull request by ISauve**
### What does this PR do?

Adds the userspace side of HTTP monitoring in the system probe. The added components are responsible for processing HTTP transactions coming from ebpf, storing http transaction information in the appropriate ConnectionStats objects, and encoding the http data in order to be sent to the process agent & then along to the backend.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan
In the agent VM:
1) Enable HTTP monitoring in `system_probe.yaml`
    ```
    network_config:
          enable_http_monitoring: true
    ```
2) Build & run the system probe
    ```
    inv -e system-probe.build --bundle-ebpf && sudo ./bin/system-probe/system-probe
    ```
3) Make some http requests (must be http, not https)
     - a good endpoint to make requests to is `http://httpbin.org/status/{code}`, which generates a response with the given code
4) Fetch the connection stats from the system probe
    ```
    curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections?client\_id=1 | jq .
    ```
5) Find the connection stats corresponding to the http requests you made & verify that they were recorded in the "httpStatsByPath" field

### CODE REVIEW

File: pkg/network/http/http_statkeeper.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+)
+
+type Key struct {
+	SourceIP   util.Address
+	DestIP     util.Address
+	SourcePort uint16
+	DestPort   uint16
+}
```
Comment (p-lambert) [inline pkg/network/http/http_statkeeper.go:16]: Just something to think about: For HTTP requests without keep alives each (src,dst) will map to a single request/path

File: pkg/network/http/http_statkeeper.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+package http
+
+import (
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+)
+
+type Key struct {
+	SourceIP   util.Address
```
Comment (p-lambert) [inline pkg/network/http/http_statkeeper.go:12]: It might be worth it indexing the data using the high/low int64 representation (and doing conversions when you fetch the data), but let's profile first :)

File: pkg/network/http/monitor.go [hunk 3]
```diff
@@ -29,20 +28,11 @@ type Monitor struct {
...
-	filter, _ := m.GetProbe(manager.ProbeIdentificationPair{Section: string(probes.SocketHTTPFilter)})
-	if filter == nil {
-		return nil, fmt.Errorf("error retrieving socket filter")
-	}
-
-	closeFilterFn, err := network.HeadlessSocketFilter(procRoot, filter)
-	if err != nil {
-		return nil, fmt.Errorf("error enabling HTTP traffic inspection: %s", err)
-	}
-
```
Comment (p-lambert) [inline pkg/network/http/monitor.go:45]: Curious, why did you change this?

**Review by p-lambert — COMMENTED** [review]
Gave a first pass on this. Overall I think it looks great!
I left a few comments mostly regarding performance concerns/optimizations that we'll likely need to consider at some point, but let's profile first :)

File: pkg/network/http/monitor.go (hunk 3 above)
Comment (ISauve) [reply pkg/network/http/monitor.go]: It was because there was an import cycle between the network & http packages - but this doesn't fix the tests, so I removed this change and instead made a new `filter` package to put `HeadlessSocketFilter` in

File: pkg/network/http/http_statkeeper.go (hunk 1 above)
Comment (ISauve) [reply pkg/network/http/http_statkeeper.go]: Very good point - I ended up taking `SourcePort` out of the key for this reason

**Comment by ISauve** [comment]
Note: currently this is blocked on https://github.com/DataDog/agent-payload/pull/78

File: go.sum [hunk 4]
```diff
@@ -85,6 +85,25 @@ github.com/DATA-DOG/go-sqlmock v1.3.3/go.mod h1:f/Ixk793poVmq4qj/V1dPUg2JEAKC73Q
...
+github.com/DataDog/agent-payload v4.53.1-0.20210113165348-51de575020e8+incompatible h1:gXJcHEIpjmrM0WX7oYUjDkZ05z4bqW+vuIvMKInp6fo=
+github.com/DataDog/agent-payload v4.53.1-0.20210113165348-51de575020e8+incompatible/go.mod h1:/2RW4IC/2z54jtB6RLgq5UtVI1TsX0joDRjKbkLT+mk=
+github.com/DataDog/agent-payload v4.53.1-0.20210113193053-81c4274e5061+incompatible h1:Qm+zDmtCKUnGH5AkmgyZv2ankfXVzR4Md9Yp5z5n4W4=
+github.com/DataDog/agent-payload v4.53.1-0.20210113193053-81c4274e5061+incompatible/go.mod h1:/2RW4IC/2z54jtB6RLgq5UtVI1TsX0joDRjKbkLT+mk=
+github.com/DataDog/agent-payload v4.53.1-0.20210114152954-be676bf33059+incompatible h1:RbpQgkQIFyCRi3RMj5PCNmbb/g0u7AW2fDyKltOYWEs=
+github.com/DataDog/agent-payload v4.53.1-0.20210114152954-be676bf33059+incompatible/go.mod h1:/2RW4IC/2z54jtB6RLgq5UtVI1TsX0joDRjKbkLT+mk=
+github.com/DataDog/agent-payload v4.53.1-0.20210114154107-05ad7aa9158d+incompatible h1:K62xDSk0vuDe64XpB1UdGJlT0X4YfrYyyBVYflqWnUM=
+github.com/DataDog/agent-payload v4.53.1-0.20210114154107-05ad7aa9158d+incompatible/go.mod h1:/2RW4IC/2z54jtB6RLgq5UtVI1TsX0joDRjKbkLT+mk=
```
Comment (sgnn7) [inline go.sum:106]: A small nit but I think the `.sum` file only needs one version's hashes so I think you can delete the v4.49 and 4.50.1 hashes from this file

File: pkg/ebpf/bytecode/bindata/bindataOffsetguessO.go [hunk 5]
```diff
@@ -28,7 +28,7 @@ func bindataOffsetguessO() (*asset, error) {
 		name:        "offset-guess.o",
 		size:        12360,
 		md5checksum: "",
-		mode:        os.FileMode(420),
+		mode:        os.FileMode(436),
```
Comment (sgnn7) [inline pkg/ebpf/bytecode/bindata/bindataOffsetguessO.go:31]: Tiny nit: I don't know if these files are generated but it would be helpful to have a comment about what all these mode values are in octal right next to their use for easier visual parsing:
```suggestion
		mode:        os.FileMode(436),  // Octal: 0664
```

Comment (ISauve) [reply pkg/ebpf/bytecode/bindata/bindataOffsetguessO.go:31]: They are generated, so unfortunately that would get overwritten

File: pkg/network/http/http_statkeeper.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	"strings"
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+)
+
+type Key struct {
+	SourceIP util.Address
+	DestIP   util.Address
+	DestPort uint16
```
Comment (p-lambert) [inline pkg/network/http/http_statkeeper.go:15]: As mentioned in our chat yesterday, if you disregard SourcePort you're in effect losing container resolution

File: pkg/network/http/model.go [hunk 7]
```diff
@@ -82,6 +90,36 @@ func (tx *httpTX) Method() string {
...
+func (tx *httpTX) SourcePort() uint16 {
+	return uint16(tx.tup.sport)
+}
+
+func (tx *httpTX) DestPort() uint16 {
+	return uint16(tx.tup.dport)
+}
+
+func (tx *httpTX) RequestLatency() time.Duration {
+	return time.Duration(int64((tx.response_last_seen-tx.request_started)/(1000000))) * time.Millisecond
```
[the remark this answers is not in the document]
Comment (ISauve) [reply pkg/network/http/model.go]: Yeah this was an error - I wanted to convert from ns to ms and thought I needed `* time.Millisecond` at the end for time.Duration to know the value I was giving it was in ms 😅  I ended up switching to use a float representation anyways though

File: pkg/network/http/request_stats.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	"time"
+
+	model "github.com/DataDog/agent-payload/process"
+	"github.com/DataDog/sketches-go/ddsketch"
+)
+
+// RequestStats stores stats for HTTP requests to a particular path, organized by the class
+// of the response code (1XX, 2XX, 3XX, 4XX, 5XX)
+type RequestStats [5]struct {
+	count     int
```
[the remark this answers is not in the document]
Comment (ISauve) [reply pkg/network/http/request_stats.go:19]: It does, but technically every time we add a value to the sketch it's possible for the sketch to not accept that value (ie if the value is outside the range that is tracked by the sketch), so I figured keeping a separate count which we know captures 100% of processed transactions is better

File: pkg/network/http/request_stats.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	latencies *ddsketch.DDSketch
+}
+
+func (r *RequestStats) addRequest(statusClass int, latency time.Duration) {
+	i := statusClass/100 - 1
+	r[i].count++
+
+	if r[i].latencies == nil {
+		relativeAccuracy := 0.01
```
[the remark this answers is not in the document]
Comment (ISauve) [reply pkg/network/http/request_stats.go]: 👍  good idea!

File: pkg/network/http/http_statkeeper.go (hunk 6 above)
Comment (ISauve) [reply pkg/network/http/http_statkeeper.go]: oops yes I did make that change but just forget to add it as part of this commit 😬   it's there now!

File: pkg/network/http/request_stats.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (p-lambert) [reply pkg/network/http/request_stats.go:17]: I guess we can keep it for now then! Could you just please add a small comment explaining your motivation?

[the remark this answers is not in the document]
Comment (ISauve) [reply pkg/network/http/request_stats.go:17]: Definitely :)

**Review by p-lambert — APPROVED** [review]
Awesome work!
The only thing that this PR is missing in my opinion is a proper integration test.
I think it would be great if you could add one, or modify `TestEnableHTTPMonitoring` to do a http request, call `GetActiveConnections` and see that you have the associated HTTP stats.

File: pkg/network/encoding/encoding_test.go [hunk 10]
```diff
@@ -89,6 +95,33 @@ func TestSerialization(t *testing.T) {
 						DnsCountByRcode:      map[uint32]uint32{0: 1},
 					},
 				},
+
+				HttpStatsByPath: map[string]*model.HTTPStats{
+					"/testpath": {
+						StatsByResponseStatus: []*model.HTTPStats_Data{
+							{
```
[the remark this answers is not in the document]
Comment (ISauve) [reply pkg/network/encoding/encoding_test.go:102]: Great idea 👍
